src/main/java/com/ebivariation/contigalias/service/AssemblyService.java

```diff
@@ -16,36 +16,152 @@
+
+    private final AssemblyDataSource dataSource;
+
+    private final ExecutorService executor = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
 
     @Autowired
-    public AssemblyService(@Qualifier("ftpDao") AssemblyDao assemblyDao) {
-        this.assemblyDao = assemblyDao;
+    public AssemblyService(AssemblyRepository repository, @Qualifier("NCBIDataSource") AssemblyDataSource dataSource) {
+        this.repository = repository;
+        this.dataSource = dataSource;
     }
 
-    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) throws IOException {
-        Optional<AssemblyEntity> assembly = assemblyDao.getAssemblyByAccession(accession);
-        assembly.ifPresent(asm -> {
+    public Optional<AssemblyEntity> getAssemblyOrFetchByAccession(String accession) throws IOException {
+        Optional<AssemblyEntity> assembly = getAssemblyByAccession(accession);
+        if (assembly.isPresent()) {
+            return assembly;
+        }
+        Optional<AssemblyEntity> fetchAssembly = fetchAndInsertAssembly(accession);
+        if (fetchAssembly.isPresent()) {
+            AssemblyEntity asm = fetchAssembly.get();
             List<ChromosomeEntity> chromosomes = asm.getChromosomes();
             if (chromosomes != null) {
                 chromosomes.forEach(chr -> chr.setAssembly(null));
             }
-        });
+            // Setting this to null to remove inconsistency in results
+            // When fetching from db bring no results, chromosomes is an empty list
+            // whereas here chromosomes is value
+            else {
+                asm.setChromosomes(new LinkedList<>());
+            }
+            return fetchAssembly;
+        }
+        return Optional.empty();
+    }
+
+    public Optional<AssemblyEntity> fetchAndInsertAssembly(String accession) throws IOException {
+        if (repository.findAssemblyEntityByAccession(accession).isPresent()) {
+            throw duplicateAssemblyInsertionException(accession);
+        }
+        Optional<AssemblyEntity> fetchAssembly = dataSource.getAssemblyByAccession(accession);
+        fetchAssembly.ifPresent(this::insertAssembly);
+        return fetchAssembly;
+    }
+
+    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) {
+        Optional<AssemblyEntity> assembly = repository.findAssemblyEntityByAccession(accession);
+        assembly.ifPresent(this::stripAssemblyFromChromosomes);
         return assembly;
     }
 
+    private void stripAssemblyFromChromosomes(AssemblyEntity assembly) {
+        List<ChromosomeEntity> chromosomes = assembly.getChromosomes();
+        if (chromosomes != null && chromosomes.size() > 0) {
+            chromosomes.forEach(chr -> chr.setAssembly(null));
+        }
+    }
+
+    public void insertAssembly(AssemblyEntity entity) {
+        setCacheSizeLimit();
+
+        if (isEntityPresent(entity)) {
+            throw duplicateAssemblyInsertionException(entity);
```

Comment:
when you give an error that something is already present, it's more useful to show what was already present (not what you are providing. Or even better, show both.